subprojects/core/src/main/java/org/gradle/api/internal/classpath/DefaultModuleRegistry.java

```diff
@@ -196,27 +196,48 @@ private Module module(String moduleName, Properties properties, Set<File> implem
     }
 
     private void findImplementationClasspath(String name, Collection<File> implementationClasspath) {
+        List<String> suffixes = getClasspathSuffixes(name);
+        for (File file : classpath) {
+            if (file.isDirectory()) {
+                for (String suffix : suffixes) {
+                    if (file.getAbsolutePath().endsWith(suffix)) {
+                        implementationClasspath.add(file);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Provides the locations where the classes and resources of a Gradle module can be found
+     * when running in embedded mode from the IDE.
+     *
+     * <ul>
+     * <li>In Eclipse, they are in the bin/ folder.</li>
+     * <li>In IDEA, they are in a folder derived from their module name.</li>
+     * Due to de-duplication with names in the buildSrc project, that module name can start with "gradle-"
+     * </ul>
+     * <li>In both cases we also include the static and generated resources of the project.</li>
+     */
+    private List<String> getClasspathSuffixes(String name) {
         List<String> suffixes = new ArrayList<String>();
         Matcher matcher = Pattern.compile("gradle-(.+)").matcher(name);
         matcher.matches();
         String projectDirName = matcher.group(1);
         String projectName = toCamelCase(projectDirName);
+
         suffixes.add(("/out/production/" + projectName).replace('/', File.separatorChar));
+        suffixes.add(("/out/production/gradle-" + projectName).replace('/', File.separatorChar));
+
         suffixes.add(("/" + projectDirName + "/bin").replace('/', File.separatorChar));
+
         suffixes.add(("/" + projectDirName + "/src/main/resources").replace('/', File.separatorChar));
-        suffixes.add(("/" + projectDirName + "/build/classes/main").replace('/', File.separatorChar));
+        suffixes.add(("/" + projectDirName + "/build/classes/java/main").replace('/', File.separatorChar));
```

Comment:
This was broken since Gradle 4.0 and is probably responsible for most Slack messages along the lines of "Has anyone seen this classloading error before when in embedded mode?"